### PR TITLE
Array Combinators

### PR DESCRIPTION
**Pull request by willow-ahrens**
This PR adds an interface for array combinators (wrapper arrays that modify the behavior of their parents). It also re-implements index modifiers using array combinators.

Several aspects of the lowering machinery have been simplified or removed. In particular, select, instantiate, unfurl are no longer requested using styles. This PR implements https://github.com/willow-ahrens/Finch.jl/issues/223, makes great strides towards https://github.com/willow-ahrens/Finch.jl/issues/43, as well as simplifying the behavior of dimensionalization quite a bit.

index expressions like `i + 1` are now converted to wrapper arrays in wrapperize.jl. Afterwards, the program is dimensionalized, and only unmodified indices participate in dimensionalization.

Also, we have added first-class scalar variables, and modified the semantics of sequence statements to restrict the scope during lowering.

There are a few remaining tasks:

- [x] Reimplement permissive, offset, protocolized
- [x] Document RootArray and tensor scoping/dimensionalization
- [x] Implement ToplitzArray
- [x] Implement WindowArray
- [x] Document combinator interface
- [x] rm forget node
- [x] rm updater mode
- [x] rm unfurlstyle
- [x] rm dimensionalizestyle

### CODE REVIEW

File: docs/src/semantics.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+variable named `_` or `:`. 
+
+Similarly, tensor declarations also set the dimensions of a tensor. Accessing a tensor with an unmodified loop index
+"hints" that the tensor axis should have the same dimension as the corresponding loop. Finch will automatically dimensionalize declarations based on all updates up to the first read.  
+
+# Array Combinators
+
+Finch includes several array combinators that modify the behavior of arrays. For
+example, the `OffsetArray` type wraps an existing array, but shifts its
+indices. The `PermissiveArray` type wraps an existing array, but allows
```
Comment (wraith1995) [inline docs/src/semantics.md:101]: allows them and produces what?

File: src/FinchNotation/instances.jl [hunk 2]
```diff
@@ -129,51 +121,45 @@ Base.show(io::IO, node::AccessInstance) = print(io, "access_instance(", node.tns
...
-Base.:(==)(a::VariableInstance, b::VariableInstance) = false
-Base.:(==)(a::VariableInstance{tag}, b::VariableInstance{tag}) where {tag} = a.tns == b.tns
-
-@inline variable_instance(tag, tns) = VariableInstance{tag, typeof(tns)}(tns)
-@inline variable_instance(tag, tns::IndexInstance) = tns #TODO this should be syntactic
 
-Base.show(io::IO, node::VariableInstance{tag}) where {tag} = print(io, "variable_instance(:", tag, ", ", tag, ")")
 
-struct ReaderInstance end
+struct TagInstance{tag, Tns} <: FinchNodeInstance
```
[the remark this answers is not in the document]
Comment (wraith1995) [reply src/FinchNotation/instances.jl:126]: (Distinguished from `VariableInstance`?)

File: src/tensors/wrapperize.jl [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+"""
+    get_wrapper_rules(alg, shash)
+
+Return the wrapperizing rule set for Finch, which converts expressions like `A[i
++ 1]` to array combinator expressions like `OffsetArray(A, (1,))`. The rules have
+access to the algebra `alg` and the depth lookup `depth`` One can dispatch on
+the `alg` trait to specialize the rule set for different algebras. These rules run
+after simplification so one can expect constants to be folded.
+"""
+function get_wrapper_rules(alg, depth, ctx)
```
Comment (wraith1995) [inline src/tensors/wrapperize.jl:55]: Is there any reason to be concerned about rewrite rule ordering or anything of the like here? I suppose if it causes problems one could just use the wrappers manually...

**Review by nullplay — APPROVED** [review]
Looks good to me.

File: src/symbolic/simplify_program.jl
[the remark this answers is not in the document]
Comment (willow-ahrens) [reply src/symbolic/simplify_program.jl:1]: I never used this feature, so I cut it out.

**Pull request merged**